lib/mu-msg.c

```diff
@@ -520,6 +520,76 @@ get_body (MuMsg *self, MuMsgOptions opts, gboolean want_html)
 		return g_string_free (bdata.gstr, FALSE);
 }
 
+
+struct _ContentTypeData {
+	const GMimeContentType *ctype;
+	gboolean want_html;
+};
+typedef struct _ContentTypeData ContentTypeData;
+
+
+static void
+find_content_type (MuMsg *msg, MuMsgPart *mpart, ContentTypeData *cdata)
+{
+	if (!GMIME_IS_PART(mpart->data))
+		return;
+
+	/* text-like attachments are included when in text-mode */
+
+	GMimePart *wanted = NULL;
+
+	if (!cdata->want_html &&
+	    (mpart->part_type & MU_MSG_PART_TYPE_TEXT_PLAIN))
+		wanted = mpart->data;
+	else if (!(mpart->part_type & MU_MSG_PART_TYPE_ATTACHMENT) &&
+		 cdata->want_html &&
+		 (mpart->part_type & MU_MSG_PART_TYPE_TEXT_HTML))
+		wanted = mpart->data;
+	if (wanted)
+		cdata->ctype = g_mime_object_get_content_type (wanted);
+}
+
+
+static const GSList*
+get_content_type_parameters (MuMsg *self, MuMsgOptions opts, gboolean want_html)
+{
+	ContentTypeData cdata;
+	cdata.want_html = want_html;
+	cdata.ctype = NULL;
+
+	mu_msg_part_foreach (self, opts,
+			     (MuMsgPartForeachFunc)find_content_type,
+			     &cdata);
+
+	if (cdata.ctype) {
+		const GSList *paramlist;
+		const GMimeParam *param;
+
+		paramlist = NULL;
+		param = g_mime_content_type_get_params (cdata.ctype);
+
+		for (; param; param = param->next) {
+			const GSList *item;
+
+			item = g_slist_prepend (NULL, param->value);
+			item = g_slist_prepend (item, param->name);
+			paramlist = g_slist_prepend (paramlist, item);
+		}
+
+		return g_slist_reverse (paramlist);
```

Comment:
I think this is a memory leak. I tried using `free_later_lst`, but this made the tests crash. I *think* this is because `mu_msg_destroy` assumes that all entries in `_free_later_lst` are lists of strings rather than lists of tuples (lists). Do you think it's best to flatten the list, or add a new `_free_later_tuples` member to `MuMsg` struct?